### PR TITLE
MDEV-32974 : Member fails to join due to old seqno in GTID

### PR DESCRIPTION
**Pull request by janlindstrom**
Before MDEV-15158, wsrep xid information was stored in only one place: in the TRX_SYS page. Starting with 10.3, it is not stored there but in the rollback segment header pages, and the latest one is what matters. MDEV-19229 allows the undo tablespaces to be rebuilt when innodb_undo_tablespaces is changed on startup. Previously it was not possible to change that parameter.

These changes caused the fact that rollback segment header pages could contain several wsrep xid's storead and when undo tablespaces were rebuilt there was a effort to restore wsrep xid back to rollback segment header page but because there was several of them the latest wsrep xid was overwritten with more older one.

trx_rseg_read_wsrep_checkpoint
	We should overwrite current recovered wsrep xid only if it's seqno
	is larger than currently recovered wsrep xid.

### CODE REVIEW

File: storage/innobase/trx/trx0rseg.cc [hunk 1]
```diff
@@ -187,16 +187,29 @@ bool trx_rseg_read_wsrep_checkpoint(const trx_rsegf_t* rseg_header, XID& xid)
...
 
-	xid.bqual_length = static_cast<int>(
+	temp_xid.bqual_length = static_cast<int>(
 		mach_read_from_4(
 			TRX_RSEG_WSREP_XID_BQUAL_LEN + rseg_header));
 
-	memcpy(xid.data, TRX_RSEG_WSREP_XID_DATA + rseg_header, XIDDATASIZE);
+	memcpy(temp_xid.data, TRX_RSEG_WSREP_XID_DATA + rseg_header, XIDDATASIZE);
+
+	long long tmp_seqno = wsrep_xid_seqno(&temp_xid);
```
Comment (temeo) [inline storage/innobase/trx/trx0rseg.cc:202]: Could be `const long long`.

File: storage/innobase/trx/trx0rseg.cc [hunk 2]
```diff
@@ -187,16 +187,29 @@ bool trx_rseg_read_wsrep_checkpoint(const trx_rsegf_t* rseg_header, XID& xid)
...
+
+	/* Set new xid only if its seqno is larger than current
+	   seqno. */
+	if (tmp_seqno > (long long)wsrep_xid_seqno(&xid))
```
Comment (temeo) [inline storage/innobase/trx/trx0rseg.cc:206]: Cast to `long long` should not be necessary. The return value of `wsrep_xid_seqno()` is `long long`.

Should we also check if `xid` is wsrep XID? e.g.

```if (!wsrep_is_wsrep_xid(&xid) || tmp_seqno > wsrep_xid_seqno(&xid))```

**Review by temeo — CHANGES_REQUESTED** [review]
In addition to requested changes, please change also the function documentation:
* `xid` is now `[in|out]` parameter
* `xid` is not updated if it already contains a wsrep XID with higher sequence number

File: storage/innobase/trx/trx0rseg.cc (hunk 2 above)
Comment (sciascid) [reply storage/innobase/trx/trx0rseg.cc:206]: In my opinion, this function should just read and return whatever xid it finds in rollback segment header. The comparison to some other xid, or any other logic, should be left to the caller.

File: storage/innobase/trx/trx0rseg.cc (hunk 1 above)
Comment (janlindstrom) [reply storage/innobase/trx/trx0rseg.cc]: Fixed differently.

**Pull request closed without merging**